common/unzip.cpp

```diff
@@ -1031,382 +906,58 @@ static int unzlocal_CheckCurrentFileCoherencyHeader(unz_s* s, uInt* piSizeVar,
-		return 1;
-	else
-		return 0;
-}
-
-
-
-/*
-  Read extra field from the current file (opened by unzOpenCurrentFile)
-  This is the local-header version of the extra field (sometimes, there is
-	more info in the local-header version than in the central-header)
-
-  if buf==NULL, it return the size of the local extra field that can be read
-
-  if buf!=NULL, len is the size of the buffer, the extra header is copied in
-	buf.
-  the return value is the number of bytes copied in buf, or (if <0)
-	the error code
-*/
-int unzGetLocalExtrafield(unzFile file, voidp buf, unsigned len) {
-	unz_s* s;
-	file_in_zip_read_info_s* pfile_in_zip_read_info;
-	uInt read_now;
-	uLong size_to_read;
-
-	if (file==nullptr)
-		return UNZ_PARAMERROR;
-	s=(unz_s*)file;
-	pfile_in_zip_read_info=s->pfile_in_zip_read;
-
-	if (pfile_in_zip_read_info==nullptr)
-		return UNZ_PARAMERROR;
-
-	size_to_read = (pfile_in_zip_read_info->size_local_extrafield -
-				pfile_in_zip_read_info->pos_local_extrafield);
-
-	if (buf==nullptr)
-		return (int)size_to_read;
-
-	if (len>size_to_read)
-		read_now = (uInt)size_to_read;
-	else
-		read_now = (uInt)len;
-
-	if (read_now==0)
-		return 0;
-
-	pfile_in_zip_read_info->_stream->seek(pfile_in_zip_read_info->offset_local_extrafield +
-			  pfile_in_zip_read_info->pos_local_extrafield,SEEK_SET);
-	if (pfile_in_zip_read_info->_stream->err())
-		return UNZ_ERRNO;
-
-	if (pfile_in_zip_read_info->_stream->read(buf,(uInt)size_to_read)!=size_to_read)
-		return UNZ_ERRNO;
-
-	return (int)read_now;
-}
-
-/*
-  Close the file in zip opened with unzipOpenCurrentFile
-  Return UNZ_CRCERROR if all the file was read but the CRC is not good
-*/
-int unzCloseCurrentFile(unzFile file) {
-	int err=UNZ_OK;
-
-	unz_s* s;
-	file_in_zip_read_info_s* pfile_in_zip_read_info;
-	if (file == nullptr)
-		return UNZ_PARAMERROR;
-	s = (unz_s*)file;
-	pfile_in_zip_read_info = s->pfile_in_zip_read;
-
-	if (pfile_in_zip_read_info == nullptr)
-		return UNZ_PARAMERROR;
-
-
-#ifdef USE_ZLIB
-	// Only verify crc32_data when zlib is linked in, because otherwise crc32() is
-	// not defined.
-	if (pfile_in_zip_read_info->rest_read_uncompressed == 0) {
-		if (pfile_in_zip_read_info->crc32_data != pfile_in_zip_read_info->crc32_wait)
-			err=UNZ_CRCERROR;
+	uint32 crc32_wait = s->cur_file_info.crc;
+	
+	byte *compressedBuffer = new byte[s->cur_file_info.compressed_size];
+	s->_stream->seek(s->cur_file_info_internal.offset_curfile + SIZEZIPLOCALHEADER + iSizeVar);
+	s->_stream->read(compressedBuffer, s->cur_file_info.compressed_size);
+	byte *uncompressedBuffer = nullptr;
+
+	switch (s->cur_file_info.compression_method) {
+	case 0: // Store
+		uncompressedBuffer = compressedBuffer;
+		break;
+	case Z_DEFLATED:
+		uncompressedBuffer = new byte[s->cur_file_info.uncompressed_size];
+		Common::GzioReadStream::deflateDecompress(uncompressedBuffer, s->cur_file_info.uncompressed_size, compressedBuffer, s->cur_file_info.compressed_size);
```

Comment:
Would it be possible to create a `GzioReadStream` object here and return that instead of decompressing the whole file up front?

Reply:
Yes, it's possible. But:
* It has different complexity for seeking. Seeking forward is cheap but seeking backwards may need to rerun the decompressor.
* It doesn't check CRC as it's impossible to check it without full read

So I decided to keep those non-transparent changes out of this PR

Reply:
Also the only thread-safe way to read from the same stream is to use MemoryReadStream in some form